account_bank_statement_import_paypal/models/account_bank_statement_import_paypal.py

```diff
@@ -0,0 +1,265 @@
+# -*- coding: utf-8 -*-
+# Copyright 2014-2017 Akretion (http://www.akretion.com).
+# @author Alexis de Lattre <alexis.delattre@akretion.com>
+# @author Sébastien BEAU <sebastien.beau@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+from datetime import datetime
+from openerp import models, fields, api, _
+from openerp.exceptions import UserError
+import re
+from cStringIO import StringIO
+_logger = logging.getLogger(__name__)
+
+try:
+    import unicodecsv
+except (ImportError, IOError) as err:
+    _logger.debug(err)
+
+# Paypal header depend of the country the order are the same but the
+# value are translated. You can add you header here
+
+HEADERS = [
+    '"Date","Heure","Fuseau horaire","Description","Devise"',  # French
+    '"Date","Time","Time Zone","Description","Currency"',  # English
+    ]
+
+
+class AccountBankStatementImport(models.TransientModel):
+    _inherit = 'account.bank.statement.import'
+
+    @api.model
+    def _get_paypal_encoding(self):
+        return 'utf-8'
+
+    @api.model
+    def _get_paypal_date_format(self):
+        '''This method is designed to be inherited'''
+        return '%d/%m/%Y'
+
+    @api.model
+    def _paypal_convert_amount(self, amount_str):
+        '''This method is designed to be inherited'''
+        valstr = re.sub(r'[^\d,.-]', '', amount_str)
+        valstrdot = valstr.replace('.', '')
+        valstrdot = valstrdot.replace(',', '.')
+        return float(valstrdot)
+
+    @api.model
+    def _check_paypal(self, data_file):
+        for header in HEADERS:
+            if data_file.strip().startswith(header):
+                return True
+        return False
+
+    def _convert_paypal_line_to_dict(self, idx, line):
+        date_dt = datetime.strptime(line[0], self._get_paypal_date_format())
+        rline = {
+            'date': fields.Date.to_string(date_dt),
+            'time': line[1],
```

Comment:
I had a bad experience where the client could download csv files whose number and order of columns is not always the same. It was a bug source (ex balance was not in column 8 but 10). One solution can be the use of csv.DictReader if we can detect csv lang (fr, en, ...) or add lang in import wizard. File name will match with HEADERS

Reply:
We do not use the same csv file as before. Indeed the previous version of this module was using the configurable export (where you can specify if you want to use your separtor, you can filter the transaction you want). Now we are using the official statement report (not configuration is possible). So I do not think we will have such issue

Reply:
OK, but the "HEADERS = [ '"Date","Heure","Fuseau horaire","Description","Devise"', .. # French" Contain only 5 columns but here we have at least 18 columns ('origin_transaction_id': line[17]). 

I propose either to put all the columns of the official statement report in the HEADER or to add a comment of the correspondence between the columns used and it index (ex: line [17] # col "transaction origin". Paypal can change the format, or a user may be mistaken and use a configurable export format. In which case it is extremely difficult to debug if we do not know the names of the columns used.

Reply:
I have set the full header to avoid the issue you describe

Reply:
:+1:  Thanks